pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp

```diff
@@ -265,8 +267,18 @@ class TNCudaBase : public TensornetBase<PrecisionT, Derived> {
             "Unsupported gate: MPS method only supports 1, 2-wires gates");
 
         auto &&par = (params.empty()) ? std::vector<PrecisionT>{0.0} : params;
-        DataBuffer<PrecisionT, int> dummy_device_data(
-            Pennylane::Util::exp2(wires.size()), getDevTag());
+
+        int64_t dummy_id = gate_ids_.empty() ? 1 : *gate_ids_.rbegin() + 1;
+
+        if (!gate_matrix.empty()) {
+            auto gate_key = std::make_pair(opName, par);
+            std::vector<CFP_t> matrix_cu =
+                cuUtil::complexToCu<ComplexT>(gate_matrix);
+            gate_cache_->add_gate(dummy_id, gate_key, matrix_cu, adjoint);
+        } else {
+            gate_cache_->add_gate(dummy_id, opName, par, adjoint);
+        }
```

Comment:
If the `if` statement is hitting several times, we can optimize the branch prediction with `[[likely]]` and ``[[unlikely]]``  [ref](https://en.cppreference.com/w/cpp/language/attributes/likely)

Reply:
This is fantastic. Thank you for sharing this knowledge, @LuisAlfredoNu.

Reply:
Thanks @LuisAlfredoNu . Is this keyword portable for `MSVC`?

Reply:
@multiphaseCFD This is part of the `std` 20 I can guess that it is portable to `MSVC`. In this [ref](https://stackoverflow.com/questions/1440570/likely-unlikely-equivalent-for-msvc) mention that is included. However, we need to do test. :thinking: